### PR TITLE
Switched to lib with added generic functionality

### PR DESCRIPTION
**Pull request by BlackPhlox**
Switched impl. to the library and added the main as an example instead. Added more generic methods though the API is not perfect yet (imo). The cached CI part for the datadump is missing and is properly something I need help with as I am not that familiar with Github CI as opposed to Azure Pipeline

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,378 @@
...
+    info!("Reading file");
+    Ok(std::fs::read_to_string(path).map(drop)?)
+}
+
+#[instrument]
+fn read_config() -> Result<(), Report> {
+    read_file("fake_file")
+        .wrap_err("Unable to read config")
+        .suggestion("Double-check that the file exist in the current path")
+}
```
Comment (alyti) [inline src/lib.rs]: These were mostly from eyre and could be yeeted.

File: src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,378 @@
...
+    let filter_layer = EnvFilter::try_from_default_env()
+        .or_else(|_| EnvFilter::try_new("info"))
+        .unwrap();
+
+    tracing_subscriber::registry()
+        .with(filter_layer)
+        .with(fmt_layer)
+        .with(ErrorLayer::default())
+        .init();
+}
```
Comment (alyti) [inline src/lib.rs]: Tracing setup should be handled by the binary, since this is now a library this serves no purpose.

File: rustfmt.toml [hunk 3]
```diff
@@ -0,0 +1 @@
+fn_args_layout = "Compressed"
```
Comment (alyti) [inline rustfmt.toml]: What's this for?

File: Cargo.toml [hunk 4]
```diff
@@ -14,4 +14,4 @@ tracing = { version = "0.1.13", optional = true }
 tracing-subscriber = { version = "0.2.5", optional = true }
 tracing-error = { version = "0.1.2", optional = true }
 
-cratesio-dbdump-csvtab = { version = "0.2", path = "../cratesio-dbdump-csvtab"}
\ No newline at end of file
+cratesio-dbdump-csvtab = { version = "0.2.2" }#, path = "../cratesio-dbdump-csvtab"}
```
Comment (alyti) [inline Cargo.toml]: Either leave path or remove completely, commenting it out is kinda eh.

**Review by alyti — CHANGES_REQUESTED** [review]
LGTM overall, but I've left some comments here and there.

File: rustfmt.toml (hunk 3 above)
Comment (BlackPhlox) [reply rustfmt.toml]: https://rust-lang.github.io/rustfmt/?version=v1.4.36&search=fn_args_layout

Comment (BlackPhlox) [reply rustfmt.toml]: Doesn't seem to work though, so I added rustfmt::skip
If I don't I get this when running `cargo fmt`:
```rust
#[instrument]
#[allow(clippy::type_complexity)]
pub fn get_rev_dependency(
    db: &Connection, crate_name: &str, expected_crate_dep_name: &str,
) -> Result<
    Vec<
        Result<
            (
                String,
                String,
                String,
                String,
                Result<Vec<(String, String)>, cratesio_dbdump_csvtab::rusqlite::Error>,
            ),
            cratesio_dbdump_csvtab::rusqlite::Error,
        >,
    >,
    cratesio_dbdump_csvtab::rusqlite::Error,
> {
```

File: src/lib.rs (hunk 2 above)
Comment (BlackPhlox) [reply src/lib.rs]: I've added it to the example instead

**Comment by BlackPhlox** [comment]
Might I also suggest that the name of the repo be changed? I don't have an idea atm. but a more specific name related to crates.io would properly be better imo.

**Comment by BlackPhlox** [comment]
Since it's a library now, I think it would be better to make impls for CratesIODumpLoader (Maybe its own struct?) instead of importing methods. It is (imo) much more ergonomic.

**Comment by alyti** [comment]
I don't have any strong opinions about the API right now, feel free to experiment with whatever feels good.
Same goes for the crate name, could probably go for something like `cratesio-dbdump-lookup`, not sure yet.

**Comment by BlackPhlox** [comment]
I like cratesio-dbdump-lookup for its consistency, I've changed it here, feel free to change the repo name if you agree 👍

File: src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,414 @@
+use cratesio_dbdump_csvtab::rusqlite::{self, Connection, Error};
+use std::num::ParseIntError;
+use tracing::instrument;
+
+pub trait CrateFetcher {
+    fn get_crate(&self, crate_name: &str) -> Result<Option<Crate>, rusqlite::Error>;
+    fn get_keywords(&self, crate_id: &str) -> Result<Vec<String>, rusqlite::Error>;
+}
```
Comment (alyti) [inline src/lib.rs]: Since this is a trait in a library, maybe we should return our own errors instead of rusqlite ones 🤔

Comment (alyti) [reply src/lib.rs]: Notably the other lib already uses thiserror for it.

Comment (BlackPhlox) [reply src/lib.rs]: This is temp atm. plus I don't know what the best way is to do so either. Feel free to give it a go.

Comment (alyti) [reply src/lib.rs]: Alright I'll experiment with it later, prob gonna merge now and look into publishing a 0.1 version if you are okay with that.

Comment (BlackPhlox) [reply src/lib.rs]: Cool cool 👍, I think #85f97af is gonna be my last commit for today 😅

**Pull request merged**